### PR TITLE
Issue 115 Use Explicit Names in Loops

### LINKED ISSUE #115: Use Explicit Names in Loops
HTL [provides implicit variables in data-sly-list and data-sly-repeat blocks](https://github.com/Adobe-Marketing-Cloud/htl-spec/blob/master/SPECIFICATION.md#226-list). Try to avoid them and use explicit names clarifying the role of the objects instead.
```
<!--/* Bad */-->
<ul data-sly-list="${model.downloadableFiles}">
    <li>
        <a href="${item.url}">${item.name}</a>
    </li>
</ul>
 
<!--/* Good */-->
<ul data-sly-list.file="${model.downloadableFiles}">
    <li>
        <a href="${file.url}">${file.name}</a>
    </li>
</ul>
```

### PR DESCRIPTION
**Pull request by GulderBone**
fixes #115

### CODE REVIEW

File: src/main/java/com/cognifide/aemrules/htl/checks/ExplicitNamesInLoopsCheck.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import com.cognifide.aemrules.tag.Tags;
+import com.cognifide.aemrules.version.AemVersion;
+import java.util.List;
+import org.apache.sling.scripting.sightly.compiler.expression.Expression;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.plugins.html.node.Node;
+import org.sonar.plugins.html.node.TagNode;
+
+@AemVersion(
```
[the remark this answers is not in the document]
Comment (GulderBone) [reply src/main/java/com/cognifide/aemrules/htl/checks/ExplicitNamesInLoopsCheck.java]: done

File: README.md [hunk 2]
```diff
@@ -78,6 +78,10 @@ Below you will find descriptions of all rules available in **AEM Rules for Sonar
 - **AEM-19** Implicit search strategy used in Sling Query
   - `SearchStrategy` can have negative performance impact if mismatched.
   Therefore developer should always make informed decision and define strategy explicitly.
+  
+- **HTL-3** Use Explicit Names in Loops
```
Comment (michaltobiasz) [inline README.md]: Shouldn't be done like here https://github.com/Cognifide/AEM-Rules-for-SonarQube/pull/137/files#diff-04c6e90faac2675aa89e2176d2eec7d8 ?

Second question. I see duplicated id of rule:
https://github.com/Cognifide/AEM-Rules-for-SonarQube/pull/124/files#diff-04c6e90faac2675aa89e2176d2eec7d8

Please fix were needed

File: src/main/java/com/cognifide/aemrules/htl/checks/ExplicitNamesInLoopsCheck.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    static final String RULE_KEY = "HTL-3";
+
+    static final String RULE_NAME = "Explicit names in loops";
+
+    static final String RULE_MESSAGE = "Use explicit names in loops";
+
+    @Override
+    public void htlExpression(Expression expression, Node node) {
+        if (node instanceof TagNode && ((TagNode) node).getAttributes().iterator().next().getName().equals("data-sly-list")) {
```
Comment (michaltobiasz) [inline src/main/java/com/cognifide/aemrules/htl/checks/ExplicitNamesInLoopsCheck.java]: 1. I would move this condition to separate method and gave name which describes what we are checking moreover I would assign second part of condition to variable.
2. If you have constant when using equals then you should use it on left side of equals
3. What about data-sly-repeat ?

Comment (GulderBone) [reply src/main/java/com/cognifide/aemrules/htl/checks/ExplicitNamesInLoopsCheck.java]: done

File: README.md (hunk 2 above)
Comment (GulderBone) [reply README.md]: done, but should be discussed.

File: src/main/java/com/cognifide/aemrules/htl/checks/ExplicitNamesInLoopsCheck.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+    @Override
+    public void htlExpression(Expression expression, Node node) {
+        if (node instanceof TagNode && (isSlyList((TagNode)node) || isSlyRepeat((TagNode)node))) {
+            createViolation(node.getStartLinePosition(), RULE_MESSAGE);
+        }
+    }
+
+    private boolean isSlyList(TagNode node) {
+        return "data-sly-list".equals(node.getAttributes().iterator().next().getName());
```
Comment (piotr-wilczynski) [inline src/main/java/com/cognifide/aemrules/htl/checks/ExplicitNamesInLoopsCheck.java]: You are checking just one attribute here. What if the attribute is on the last place?

File: src/main/java/com/cognifide/aemrules/htl/checks/ExplicitNamesInLoopsCheck.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (GulderBone) [reply src/main/java/com/cognifide/aemrules/htl/checks/ExplicitNamesInLoopsCheck.java]: Changed to:
`return node.getAttributes().stream()
            .anyMatch(attribute -> "data-sly-list".equals(attribute.getName()));`

File: src/main/java/com/cognifide/aemrules/htl/checks/ExplicitNamesInLoopsCheck.java (hunk 4 above)
Comment (GulderBone) [reply src/main/java/com/cognifide/aemrules/htl/checks/ExplicitNamesInLoopsCheck.java]: Changed to:
```
return node.getAttributes().stream()
            .anyMatch(attribute -> "data-sly-repeat".equals(attribute.getName()));
```
For both: _data-sly-list_ and _data-sly-repeat_

**Comment by GulderBone** [comment]
https://github.com/Cognifide/AEM-Rules-for-SonarQube/pull/160
Pull request with resolved conflicts.

**Pull request closed without merging**